ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign-v4-simulation/decoded-simulation/decoded-simulation.tsx

```diff
@@ -104,19 +106,39 @@ const DecodedSimulation: React.FC<object> = () => {
   const chainId = currentConfirmation.chainId as Hex;
   const { decodingLoading, decodingData } = currentConfirmation;
 
-  const stateChangeFragment = useMemo(
-    () =>
-      (decodingData?.stateChanges ?? []).map(
-        (change: DecodingDataStateChange) => (
+  const stateChangeFragment = useMemo(() => {
+    const stateChangesGrouped: Record<string, DecodingDataStateChange[]> = (
+      decodingData?.stateChanges ?? []
+    ).reduce(
+      (
+        result: Record<string, DecodingDataStateChange[]>,
+        stateChange: DecodingDataStateChange,
+      ) => {
+        result[stateChange.changeType] = [
+          ...(result[stateChange.changeType] ?? []),
+          stateChange,
+        ];
+        return result;
+      },
+      {},
+    );
+
+    return Object.entries(stateChangesGrouped)
+      .map(([_, changeList]) =>
+        changeList.map((change: DecodingDataStateChange, index: number) => (
           <StateChangeRow
             stateChangeList={decodingData?.stateChanges ?? []}
             stateChange={change}
             chainId={chainId}
+            displayLabel={index === 0}
           />
-        ),
-      ),
-    [decodingData?.stateChanges],
-  );
+        )),
+      )
+      .reduce(
```

Comment:
Minor, rather than reducing, could we just use `flatMap` above?

Reply:
updated